File: DataStructures/Tree/AVL/python/Avl.py
import sys

class Node:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
		self.height = 1

class Avl:
	def __init__(self):
		self.root = None

	def __height(self, node):
		if node is None:
			return 0
		return node.height

	def __getBalance(self, node):
		if(node is None):
			return 0
		return self.__height(node.left) - self.__height(node.right);

	def __rightRotate(self, a):
		b = a.left
		t = b.right

		b.right = a
		a.left = t

		a.height = max(self.__height(a.left), self.__height(a.right)) + 1
		b.height = max(self.__height(b.left), self.__height(b.right)) + 1

		return b
	
	def __leftRotate(self, a):
		b = a.right
		t = b.left

		b.left = a
		a.right = t

		a.height = max(self.__height(a.left), self.__height(a.right)) + 1
		b.height = max(self.__height(b.left), self.__height(b.right)) + 1

		return b
	
	def __balancedTree(self, node, data):

		node.height = 1 + max(self.__height(node.left), self.__height(node.right))
		bal = self.__getBalance(node)

		# Right right case
		if(bal> 1 and data < node.left.data):
			return self.__rightRotate(node)

		# Left left case
		if(bal < -1 and data > node.right.data):
			return self.__leftRotate(node)

		# Left Right case
		if(bal > 1 and data > node.left.data):
			node.left =  self.__leftRotate(node.left)
			return self.__rightRotate(node)

		# Right left case
		if(bal < -1 and data < node.right.data):
			node.right = self.__rightRotate(node.right)
			return self.__leftRotate(node)

		return node

	def insert(self, root, data):
		if(root is None):
			return Node(data)
		if(data < root.data):
			root.left = self.insert(root.left, data)
		elif(data > root.data):
			root.right = self.insert(root.right, data)
		return self.__balancedTree(root, data)

	def preorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.preorder(root.left)
			self.preorder(root.right)
		
	def inorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.inorder(root.left)
			self.inorder(root.right)

	def postorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.postorder(root.left)
			self.postorder(root.right)

	def search(self, root, data):
		if(root is None):
			return False
		if(root.data == data):
			return True
		elif(root.data < data):
			return self.search(root.right, data)
		else:
			return self.search(root.left, data)
		return False

	def getInorderSuccessor(self,temp):
		while(temp and temp.left):
			temp = temp.left;
		return temp;
	
	def delete(self, root, data):
		if(root is None):
			return root
		if(data < root.data):
			root.left = self.delete(root.left, data)
		elif(data > root.data):
			root.right = self.delete(root.right, data)
		else:
			#Node with one child or no child 
			if(root.left is None):
				temp = root.right
				root = None
				return temp
			elif(root.right is None):
				temp = root.left
				root = None
				return temp
			#Node with two child
			temp = self.getInorderSuccessor(root.right)
			root.data = temp.data
			root.right = self.delete(root.right, temp.data)

		return self.__balancedTree(root, data)


if __name__ == "__main__":

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = sys.stdin.readline()
	a = sys.stdin.readline().split(" ") # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	avl = Avl();

	for i in a:
		avl.root = avl.insert(avl.root, i)

	for line in sys.stdin:

		if line[-1] == '\n':
			opt = line[:-1].split(" ")
		else:
			opt = line.split(" ")
		
		if opt[0] == "search":
			key = opt[1]
			isFound = avl.search(avl.root,int(key))
			if(isFound is True):
				print(key," is found.", end = "\n");
			else:
				print(key," is not found.", end = "\n");
		elif opt[0] == "delete":
			key = opt[1]
			avl.delete(avl.root, int(key))
		elif opt[0] == "inorder":
			print("Inorder traversal: ")
			avl.inorder(avl.root)
			print(end = "\n")
		elif opt[0] == "preorder":
			print("Preorder traversal: ")
			avl.preorder(avl.root)
			print(end = "\n")
		elif opt[0] == "postorder":
			print("Postorder traversal: ")
			avl.postorder(avl.root)
			print(end = "\n")

File: DataStructures/Tree/AVL/c/AVL.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

struct Node{
	int data;
	struct Node *left, *right;
	int height;
};

struct Node *newnode(int key){
	struct Node *node = (struct Node*)malloc(sizeof(struct Node));
	node -> data = key;
	node -> left = node -> right = NULL;
	node -> height = 1;
	return node;
}
//
int  max(int a, int b){
	return (a>b)? a:b;
}
int height(struct Node *node){
	if(node == NULL)
		return 0;
	return (node -> height);
} 
int getBalance(struct Node *node){
	if(node == NULL)
		return 0;
	return height(node->left) - height(node->right); 
}
struct Node* rightRotate(struct Node* a){
	struct Node *b = a -> left;
	struct Node *t = b -> right;

	b -> right = a;
	a -> left = t;

	a -> height = max(height(a->left), height(a->right)) + 1;
	b -> height = max(height(b->left), height(b->right)) + 1;

	return b;
}
struct Node* leftRotate(struct Node* a){
	struct Node *b = a -> right;
	struct Node *t = b -> left;

	b -> left = a;
	a -> right = t;

	a -> height = max(height(a->left), height(a->right)) + 1;
	b -> height = max(height(b->left), height(b->right)) + 1;

	return b;
}
struct Node *balancedTree(struct Node *node, int data){

	node -> height = 1 + max(height(node->left), height(node->right));
	int bal = getBalance(node);

	// Left Left case 
	if (bal> 1 && data < node->left->data) 
        return rightRotate(node); 

    // Right Right Case 
    if (bal < -1 && data > node->right->data) 
        return leftRotate(node); 
  
    // Left Right Case 
    if (bal > 1 && data > node->left->data) 
    { 
        node->left =  leftRotate(node->left); 
        return rightRotate(node); 
    } 
  
    // Right Left Case 
    if (bal < -1 && data < node->right->data) 
    { 
        node->right = rightRotate(node->right); 
        return leftRotate(node); 
    }

	return node;
}
//

struct Node *insert(struct Node *node, int data){
	if(node == NULL)
		return newnode(data);
	if(data < node->data)
		node -> left = insert(node->left, data);
	else if(data > node->data)
		node -> right = insert(node->right, data);
	else
		return node; 

	return balancedTree(node, data);
}

bool search(struct Node *root, int data){
	if(root == NULL)
		return false;
	if(root->data == data)
		return true;
	else if(root->data < data)
		return search(root->right, data);
	else
		return search(root->left, data);
	return false;
}

struct Node *getInorderSuccessor(struct Node *temp){
	while(temp && temp->left != NULL)
		temp = temp -> left;
	return temp;
}
struct Node *delete(struct Node *node, int data){
	if(node == NULL)
		return node;
	if(data < node->data )
		node -> left = delete(node->left, data);
	else if(data > node->data)
		node -> right = delete(node->right, data);
	else{
		//Node with one child or no child 
		if(node -> left == NULL){
			struct Node *temp = node -> right;
			free(node);
			return temp;
		}
		else if(node -> right == NULL){
			struct Node *temp = node -> left;
			free(node);
			return temp;
		}
		//Node with two child
		struct Node *temp = getInorderSuccessor(node->right);
		node -> data = temp -> data;
		node -> right = delete(node->right, temp -> data);
	}
	return balancedTree(node, data);
}

void preorder(struct Node* root){
	if(root){
		printf("%d ", root -> data);
		preorder(root -> left);
		preorder(root -> right);
	}
}
void inorder(struct Node* root){
	if(root){
		inorder(root -> left);
		printf("%d ", root -> data);
		inorder(root -> right);
	}
}
void postorder(struct Node* root){
	if(root){
		postorder(root -> left);
		postorder(root -> right);
		printf("%d ", root -> data);
	}
}

int main(){

	FILE* fp = freopen("../input.txt", "r", stdin);
	freopen("../output.txt", "w", stdout);

	int i,n, a[n];
	scanf("%d",&n);
	for(i=0; i<n; i++)
		scanf("%d" ,&a[i]);

	struct Node *root = NULL;
	for(i=0; i<n; i++){
		root = insert(root, a[i]);
	}

	while(!feof(fp)){

		char opt[20]; 
		fgets(opt, 20, fp); // it reads string from a file, one line at a time

		if(!strncmp(opt,"search",strlen("search"))){ 
			char choice[10];
			int key;
			sscanf(opt, "%s %d" , choice, &key); //sscanf reads formatted input from string

			bool isFound = search(root,key); 
			if(isFound)
				printf("%d is found.\n", key);
			else
				printf("%d is not found\n", key);
		}
		else if(!strncmp(opt,"delete",strlen("delete"))){
			char choice[10];
			int key;
			sscanf(opt, "%s %d" , choice, &key);

			root = delete(root,key); 
		}
		else if(!strncmp(opt,"inorder",strlen("inorder"))){
			printf("Inorder traversal: ");
			inorder(root);
			printf("\n");
		}
		else if(!strncmp(opt,"preorder",strlen("preorder"))){
			printf("Preorder traversal: ");
			preorder(root);
			printf("\n");
		}
		else if(!strncmp(opt,"postorder",strlen("postorder"))){
			printf("Postorder traversal: ");
			postorder(root);
			printf("\n");
		}
	}


	fclose(fp); 
}